core/src/main/scala/kafka/server/ClientMetricsManager.scala

```diff
@@ -17,37 +37,63 @@ import scala.jdk.CollectionConverters._
 
   def getSupportedCompressionTypes: List[java.lang.Byte] = {
     val compressionTypes = new ListBuffer[java.lang.Byte]
-    CompressionType.values.foreach(x => compressionTypes.append(x.id.toByte))
+    CompressionType.values.filter(x => x != CompressionType.NONE).foreach(x => compressionTypes.append(x.id.toByte))
     compressionTypes.toList
   }
 
+  // TODO: Needs to integrate with external plugin changes..
+  // if plugin is not configured, getMetricsSubscriptions and pushMetricSubscription needs to return errors
+  // to the client.
+  @Evolving
+  def checkCmReceiverPluginConfigured()  = {
+    if (ClientMetricsReceiverPlugin.getCmReceiver().isEmpty) {
+      throw new ClientMetricsReceiverPluginNotFoundException("Broker does not have any configured client metrics receiver plugin")
+    }
+  }
+
   def processGetTelemetrySubscriptionRequest(request: RequestChannel.Request,
                                              throttleMs: Int): GetTelemetrySubscriptionResponse = {
+    checkCmReceiverPluginConfigured()
     val subscriptionRequest = request.body[GetTelemetrySubscriptionRequest]
     val clientInfo = CmClientInformation(request, subscriptionRequest.getClientInstanceId.toString)
     _instance.processGetSubscriptionRequest(subscriptionRequest, clientInfo, throttleMs)
   }
 
+  def processPushTelemetryRequest(request: RequestChannel.Request, throttleMs: Int): PushTelemetryResponse = {
+    checkCmReceiverPluginConfigured()
+    val pushTelemetryRequest = request.body[PushTelemetryRequest]
+    val clientInfo = CmClientInformation(request, pushTelemetryRequest.getClientInstanceId.toString)
+    _instance.processPushTelemetryRequest(pushTelemetryRequest, request.context, clientInfo, throttleMs)
+  }
 }
 
 class ClientMetricsManager {
-  def getClientInstance(id: Uuid) = ClientMetricsCache.getInstance.get(id)
+
+  class ClientMetricsException(val s: String, var errorCode: Errors) extends KafkaException(s) {
+    def getErrorCode: Errors = this.errorCode
+  }
+
+  def getClientInstance(id: Uuid) : Option[CmClientInstanceState] =  {
+    Option(id) match {
+      case None | Some(Uuid.ZERO_UUID) => None
+      case _ => ClientMetricsCache.getInstance.get(id)
+    }
+  }
 
   def processGetSubscriptionRequest(subscriptionRequest: GetTelemetrySubscriptionRequest,
                                     clientInfo: CmClientInformation,
                                     throttleMs: Int): GetTelemetrySubscriptionResponse = {
-    var clientInstanceId = subscriptionRequest.getClientInstanceId
-    if (clientInstanceId == null || clientInstanceId == Uuid.ZERO_UUID) {
-      clientInstanceId = Uuid.randomUuid()
-    }
-    var clientInstance = getClientInstance(clientInstanceId)
-    if (clientInstance == null) {
-      clientInstance = createClientInstance(clientInstanceId, clientInfo)
+
+    val clientInstanceId = Option(subscriptionRequest.getClientInstanceId) match {
+      case None | Some(Uuid.ZERO_UUID) => Uuid.randomUuid()
```

Comment:
Should we add a check, how unlikely it might be, that this new uuid is not already in the cache? It would lead to hard to debug issues

Reply:
Done